memory/src/lib.rs

```diff
@@ -1,50 +1,223 @@
 }
 
 /// Raw memory just use for the storage
 /// I thinking about time series data for parallelisation
-#[derive(Debug, Clone)]
-pub struct MemoryRaw {
-    ptr: &UntypedValue,
+// #[derive(Debug, Clone)]
+struct MemoryRaw<'a> {
+    ptr: &'a UntypedValue,
     memory_raw: Vec<u8>,
     memory_trace: Vec<MemoryTrace>,
 }
 
 /// zkMemory
-pub struct Memory {
-    raw: memory_raw,
+struct Memory<'a> {
+    time_count: u64,
+    raw: MemoryRaw<'a>,
     commitment: Vec<MemoryCommitment>,
 }
 
 /// Have no idea about memory commitment but it will be great if we use verkle tree commitment
 /// To prove the memory
-struct MemoryCommitment {}
+#[derive(Clone, Debug)]
+struct MemoryCommitment {
+    bits: u64,
+}
 
 /// The interface for zkMemory
-trait MemoryInterface {
-    fn new(raw_memory: MemoryRaw) -> Self;
-    fn init(&mut self, address: u64) -> Result<MemoryCommitment, Error>;
+trait MemoryInterface<'a> {
+    fn new(raw_memory: MemoryRaw<'a>) -> Self;
+    // fn init(&mut self, address: u64) -> Result<MemoryCommitment, Error>;
     fn read(&mut self, address: u64) -> Result<MemoryCommitment, Error>;
     fn write(&mut self, address: u64, chunk: u64) -> Result<MemoryCommitment, Error>;
+    // fn extract_memory_trace(&mut self) -> Result<Vec<MemoryTrace>, Error>;
 }
 
-impl MemoryInterface for Memory {}
+// impl <'a>Memory<'a> {
+
+// }
+
+impl <'a>MemoryInterface<'a> for Memory<'a> {
+    // fn init(&mut self, address: u64) -> Result<MemoryCommitment, Error> {
+    //     Err(Error)
+    // }
+
+    fn new(raw_memory: MemoryRaw<'a>) -> Memory {
+        Memory { 
+            time_count: 0u64,
+            raw: raw_memory, 
+            commitment: Vec::new(), 
+        }
+    }
+
+    fn read(&mut self, address: u64) -> Result<MemoryCommitment, Error> {
+        // Check if the address is within the valid memory range
+        let memory_size = self.raw.memory_raw.len() as u64;
+        // Return an error if the address is invalid
+        if (address >= memory_size) || (address % CELL_SIZE != 0) {
+            self.time_count += 1;
+            return Err(Error); 
+        }
+        // Perform the read operation
+        let mut data = self.raw.memory_raw[address as usize] as u64;
+        for i in address+1..address+CELL_SIZE {
+            data *= 256;
+            data += self.raw.memory_raw[i as usize] as u64;
+        }
+        // Push a MemoryTrace to storage
+        let trace = MemoryTrace {
+            time_log: self.time_count,
+            address: UntypedValue { bits: (address) },
+            action: MemoryAction::Read,
+            value: UntypedValue { bits: data },
+        };
+
+        self.raw.memory_trace.push(trace);
+
+        // Increment the time count
+        self.time_count += 1;
+
+        // Return the MemoryCommitment with the read data
+        Ok(MemoryCommitment { bits: data as u64 })
+    }
+
+    fn write(&mut self, address: u64, chunk: u64) -> Result<MemoryCommitment, Error> {
```

Comment:
`address` used to equal to `WORD_SIZE` for example, `x86` has address space is 32 bits and also the `WORD_SIZE`. If you said address is fixed `u64` it won't work for Wasm 32 bits or EVM.